framework/src/engine/legacy/legacy_chain_handler.ts

```diff
@@ -12,41 +12,170 @@
+import { Database } from '@liskhq/lisk-db';
 import { LegacyConfig } from '../../types';
+import { Network } from '../network';
+import { getBlocksFromIdResponseSchema } from '../consensus/schema';
+import { Storage } from './storage';
+import { LegacyBlock, LegacyBlockBracket, Peer } from './types';
+import { decodeBlock, encodeBlock } from './codec';
+import { PeerNotFoundWithLegacyInfo } from './errors';
+import { validateLegacyBlock } from './validate';
 
 interface LegacyChainHandlerArgs {
 	legacyConfig: LegacyConfig;
+	network: Network;
 }
 
 interface LegacyHandlerInitArgs {
-	db: Database | InMemoryDatabase;
+	db: Database;
 }
 
 export class LegacyChainHandler {
+	private readonly _network: Network;
+	private _storage!: Storage;
 	private readonly _legacyConfig: LegacyConfig;
-	private _db!: Database | InMemoryDatabase;
+	private _previousLegacyBlock: LegacyBlock | undefined;
 
 	public constructor(args: LegacyChainHandlerArgs) {
 		this._legacyConfig = args.legacyConfig;
+		this._network = args.network;
 	}
 
-	// eslint-disable-next-line @typescript-eslint/require-await
-	public async init(args: LegacyHandlerInitArgs): Promise<void> {
-		this._db = args.db;
+	public init(args: LegacyHandlerInitArgs): void {
+		this._storage = new Storage(args.db);
 	}
 
-	// eslint-disable-next-line @typescript-eslint/require-await
-	public async syncBlocks(): Promise<void> {
-		// eslint-disable-next-line no-console
-		console.log(this._legacyConfig);
-		// eslint-disable-next-line no-console
-		console.log(this._db);
-		// TODO: implement the logic, general flow is as below
-		// 1. check config if sync legacy flag is true
-		// 2. Checks to what height blocks are present for a snapshotBlockID backwards
-		// 3. checks for all the peers node info if any peer has legacy blocks for the snapshot block ID, if not wait for X second
-		// 4. sync the blocks
-		// 5. Update node info
+	public async sync() {
+		if (!this._legacyConfig.sync) {
+			return;
+		}
+
+		for (const bracket of this._legacyConfig.brackets) {
+			const bracketInfo = await this._storage.getLegacyChainBracketInfo(bracket.snapshotBlockID);
+			if (bracket.startHeight === bracketInfo.lastHeight) {
+				continue;
+			}
+			const lastBlock = await this._storage.getBlockByHeight(bracketInfo.lastHeight);
+			const { id } = decodeBlock(lastBlock).block.header;
+			await this._trySyncBlocks(bracket, Buffer.from(id).toString('hex'));
+			await this._storage.saveBracketInfo(bracketInfo);
+		}
+
+		this._network.applyNodeInfo({
+			legacy: this._legacyConfig.brackets.map(bracket => Buffer.from(bracket.snapshotBlockID)),
+		});
+	}
+
+	private async _trySyncBlocks(bracket: LegacyBlockBracket, lastBlock: string) {
+		try {
+			await this.syncBlocks(bracket, lastBlock);
+		} catch (err) {
+			if (err instanceof PeerNotFoundWithLegacyInfo) {
+				// eslint-disable-next-line @typescript-eslint/no-misused-promises
+				setTimeout(async () => {
```

Comment:
create a private variable in this class to capture setTimeout return value and use it to clearTimeout after we finish sync.